physical/raft/snapshot.go

```diff
@@ -222,31 +310,123 @@ func (s *BoltSnapshotSink) ID() string {
 
-	return s.meta.ID
+	// Create the BoltDB file
+	dbPath := filepath.Join(path, databaseFilename)
+	boltDB, err := bolt.Open(dbPath, 0666, &bolt.Options{Timeout: 1 * time.Second})
+	if err != nil {
+		return err
+	}
+
+	// Write the snapshot metadata
+	if err := writeSnapshotMetaToDB(&s.meta, boltDB); err != nil {
+		return err
+	}
+
+	// Set the snapshot ID to the generated name.
+	s.meta.ID = name
+
+	// Create the done channel
+	s.doneWritingCh = make(chan struct{})
+
+	// Store the directories so we can commit the changes on success or abort
+	// them on failure.
+	s.dir = path
+	s.parentDir = s.store.path
+
+	// Create a pipe so we pipe writes into the below go routine.
+	reader, writer := io.Pipe()
+	s.writer = writer
+
+	// Start a go routine in charge of piping data from the snapshot's Write
+	// call to the delimtedreader and the BoltDB file.
+	go func() {
+		defer close(s.doneWritingCh)
+		defer boltDB.Close()
+
+		// The delimted reader will parse full proto messages from the snapshot
+		// data.
+		protoReader := NewDelimitedReader(reader, math.MaxInt32)
+		defer protoReader.Close()
+
+		var done bool
+		var keys int
+		entry := new(pb.StorageEntry)
+		for !done {
+			err := boltDB.Update(func(tx *bolt.Tx) error {
+				b, err := tx.CreateBucketIfNotExists(dataBucketName)
+				if err != nil {
+					return err
+				}
+
+				// Commit in batches of 50k. Bolt holds all the data in memory and
+				// doesn't split the pages until commit so we do incremental writes.
+				for i := 0; i < 50000; i++ {
+					err := protoReader.ReadMsg(entry)
+					if err != nil {
+						if err == io.EOF {
+							done = true
+							return nil
+						}
+						return err
+					}
+
+					err = b.Put([]byte(entry.Key), entry.Value)
+					if err != nil {
+						return err
+					}
+					keys += 1
+				}
+
+				return nil
+			})
+			if err != nil {
+				s.logger.Error("snapshot write: failed to write transaction", "error", err)
+				s.writeError = err
+				return
+			}
+
+			s.logger.Trace("snapshot write: writing keys", "num_written", keys)
+		}
+	}()
+
+	return nil
 }
 
-// Write is used to append to the state file. We write to the
-// buffered IO object to reduce the amount of context switches.
+// Write is used to append to the bolt file. The first call to write ensures we
+// have the file created.
 func (s *BoltSnapshotSink) Write(b []byte) (int, error) {
 	s.l.Lock()
 	defer s.l.Unlock()
 
-	// If someone is writing to this sink then we need to create a file sink to
-	// capture the data. This currently only happens when a follower is being
-	// sent a snapshot.
-	if s.fileSink == nil {
-		fileSink, err := s.store.fileSnapStore.Create(s.meta.Version, s.meta.Index, s.meta.Term, s.meta.Configuration, s.meta.ConfigurationIndex, s.trans)
-		if err != nil {
+	// If this is the first call to Write we need to setup the boltDB file and
+	// kickoff the pipeline write
+	if previouslyWritten := s.written.Swap(true); !previouslyWritten {
+		// Reap any old snapshots
```

Comment:
Is there any harm to attempting to reap old snapshots if we have not previously written?

Reply:
Yeah, we only want to reap before creating the new snapshot. Otherwise it'll be really hard to know if a snapshot is still active or being written to.